src/Npgsql/NpgsqlBinaryImporter.cs

```diff
@@ -274,48 +269,34 @@ Task Write<T>(bool async, T value, string dataTypeName, CancellationToken cancel
         if (cancellationToken.IsCancellationRequested)
             return Task.FromCanceled(cancellationToken);
 
-        var p = _params[_column];
-        if (p == null)
+        // First row, create the parameter objects
+        ref var p = ref _params[_column];
+        if (p is not NpgsqlParameter<T> typedParam)
+            p = typedParam = new NpgsqlParameter<T> { DataTypeName = dataTypeName };
+        else if (p._npgsqlDbType is not null || p._dataTypeName != dataTypeName)
         {
-            // First row, create the parameter objects
-            _params[_column] = p = typeof(T) == typeof(object)
-                ? new NpgsqlParameter()
-                : new NpgsqlParameter<T>();
+            p.ResetDbType();
```

Comment:
Instead of resetting at the right times we could also decide to throw when things don't line up across invocations.

It should be the case a binary import would have identical pg types after the first row initialized all parameters to certain pg types.

What do you guys think?

Reply:
I don't think we've ever supported changing types across rows, case in point from 7.0:

https://github.com/npgsql/npgsql/blob/148662917adc0faf10568197b9def86af711e842/src/Npgsql/NpgsqlBinaryImporter.cs#L248-L251

What's weird is that the same thing for `dataTypeName` was commented out:

https://github.com/npgsql/npgsql/blob/148662917adc0faf10568197b9def86af711e842/src/Npgsql/NpgsqlBinaryImporter.cs#L300-L303

The only issue is passing different types of values which are semantically the same (like `bool` and `bool?` with both passing `true`/`false`):

https://github.com/npgsql/npgsql/blob/148662917adc0faf10568197b9def86af711e842/src/Npgsql/NpgsqlBinaryImporter.cs#L324-L330

Though even then it's hard for me to imagine the case when it's actually possible.

IMO this is actually a design flaw. Instead of always passing `dataTypeName`/`NpgsqlDbType` they should have been used to initialize once at the start (for example, `Init(NpgsqlDbType?[]? npgsqlDbTypes, string?[]? dataTypeNames)`). But oh well, changing that might be problematic.

Anyway, I'm in favor of always throwing.

Reply:
>IMO this is actually a design flaw

Yes I agree, this should have been info passed in during `BeginBinaryImport`, we have an issue for it https://github.com/npgsql/npgsql/issues/1474


>Anyway, I'm in favor of always throwing.

Take a look again.
I've chosen to check the resolution *output*. It's obviously difficult to assure equality based on the *inputs* when we only have a clr type (i.e. `Write<T>(T value)`).